### PR TITLE
B18

### PR DESCRIPTION
**Pull request by vinniefalco**
* Increase optimization settings for MSVC builds

HTTP

* Check invariants in parse_op:
* Clean up message docs

WebSocket

* Write buffer option does not change capacity
* Close connection during async_read on close frame

Core

* Meet DynamicBuffer requirements for static_streambuf
* Fix write_frame masking and auto-fragment handling

Extras

* unit_test::suite fixes:
  - New overload of fail() specifies file and line
  - BEAST_EXPECTS only evaluates the reason string on a failure

### CODE REVIEW

**Comment by vinniefalco** [comment]
Documentation preview
http://vinniefalco.github.io/stage/beast/b18/

File: include/beast/websocket/impl/write.ipp [hunk 1]
```diff
@@ -0,0 +1,698 @@
...
+    else if(! fh.mask && ! wr_.autofrag)
+    {
+        fh.fin = fin;
+        fh.len = remain;
+        detail::fh_streambuf fh_buf;
+        detail::write<static_streambuf>(fh_buf, fh);
+        boost::asio::write(stream_,
+            buffer_cat(fh_buf.data(), buffers), ec);
+        failed_ = ec != 0;
+        if(failed_)
```
Comment (mellery451) [inline include/beast/websocket/impl/write.ipp]: this `if-return` seems pointless given the subsequent `return`

File: include/beast/websocket/impl/write.ipp [hunk 2]
```diff
@@ -0,0 +1,698 @@
...
+    different algorithms are used.
+
+    1.  autofragment: false
+        compression:  false
+
+        In the server role, this will send a single frame in one
+        system call, by concatenating the frame header and the payload.
+
+        In the client role, this will send a single frame in one system
+        calls, using the write buffer to calculate masked data.
```
Comment (vinniefalco) [inline include/beast/websocket/impl/write.ipp]: will fix

File: include/beast/websocket/impl/write.ipp (hunk 1 above)
Comment (vinniefalco) [inline include/beast/websocket/impl/write.ipp]: I prioritize preserving the idiom even for cases where it is unnecessary.

File: include/beast/core/detail/clamp.hpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace beast {
+namespace detail {
+
+template<class UInt>
+static
+std::size_t
+clamp(UInt x)
+{
+    if(x >= std::numeric_limits<std::size_t>::max())
```
Comment (vinniefalco) [inline include/beast/core/detail/clamp.hpp]: There are places in the WebSocket protocol which specify `uint64_t`. For example, the size of a frame can be 2^64-1 bytes. However, internally we use `std::size_t` to represent sizes which we might need to allocate. Clearly, an executable compiled for a 32-bit address model cannot allocate more than 2^32-1. This function exists to clamp a value that will be used for some type of computation involving a buffer. On 32 bit platforms it converts `std::uint64_t` to `std::uint32_t` by clamping. On 64 bit platforms it ends up doing nothing.

File: include/beast/websocket/impl/read.ipp [hunk 4]
```diff
@@ -541,6 +548,368 @@ upcall:
...
+            boost::asio::write(stream_, fb.data(), ec);
+            failed_ = ec != 0;
+            if(failed_)
+                return;
+        }
+        websocket_helpers::call_teardown(next_layer(), ec);
+        failed_ = ec != 0;
+        if(failed_)
+            return;
+        ec = error::failed;
```
Comment (vinniefalco) [inline include/beast/websocket/impl/read.ipp]: That won't set `failed_`. But also, it breaks this idiom:
```
failed_ = ec != 0;
if(failed_)
    return;
```

File: include/beast/core/detail/clamp.hpp (hunk 3 above)
Comment (mellery451) [inline include/beast/core/detail/clamp.hpp]: 👍

File: include/beast/websocket/impl/accept.ipp [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+                d.resp, std::move(*this));
+            return;
+
+        // sent response
+        case 1:
+            d.state = 99;
+            ec = d.final_ec;
+            if(! ec)
+                d.ws.open(detail::role_type::server);
+            break;
```
Comment (vinniefalco) [inline include/beast/websocket/impl/accept.ipp]: state always starts at 0

**Comment by bachase** [comment]
Looks good to me, but I skipped 9219ade03ec845d1669be9c4f01cf6d9ddd521cf for now.  Is that one strictly moving code around or there non-trivial changes?

**Comment by vinniefalco** [comment]
@bachase 9219ade is just splitting up what was in stream.ipp into corresponding files sorted by operation (read, write, parse, close, handshake, etc). No functional changes.

File: include/beast/websocket/impl/write.ipp (hunk 2 above)
Comment (vinniefalco) [inline include/beast/websocket/impl/write.ipp]: Actually the whole thing is removed in a later commit

File: include/beast/websocket/detail/stream_base.hpp [hunk 6]
```diff
@@ -98,14 +98,37 @@ struct stream_base
...
+        // Size of the write buffer.
+        // This gets set to the write buffer size option at the
+        // beginning of sending a message, so that the option can be
+        // changed mid-send without affecting the current message.
+        std::size_t size;
+
+        // The write buffer.
+        // The buffer is allocated or reallocated at the beginning of
+        // sending a message.
+        //
```
Comment (wilsonianb) [inline include/beast/websocket/detail/stream_base.hpp]: extra line

File: doc/websocket.qbk [hunk 7]
```diff
@@ -292,42 +291,45 @@ void echo(beast::websocket::stream<boost::asio::ip::tcp::socket>& ws)
...
+the implementation also automatically handles control frames.
 
+[heading Ping and Pong Frames]
 
-[section:pongs Pong messages]
+Ping and pong messages are control frames which may be sent at any time
+by either peer on an established WebSocket connection. They are sent
+using the functions
+    [link beast.ref.websocket__stream.ping `ping`] and
+    [link beast.ref.websocket__stream.ping `pong`].
```
Comment (wilsonianb) [inline doc/websocket.qbk]: `beast.ref.websocket__stream.ping` --> `beast.ref.websocket__stream.pong`

File: include/beast/websocket/impl/write.ipp (hunk 2 above)
Comment (vinniefalco) [inline include/beast/websocket/impl/write.ipp]: All of the implementation notes at the top of write.ipp are removed in a later branch:
https://github.com/vinniefalco/Beast/blob/pmd/include/beast/websocket/impl/write.ipp

File: include/beast/websocket/impl/accept.ipp [hunk 8]
```diff
@@ -0,0 +1,435 @@
...
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#ifndef BEAST_WEBSOCKET_IMPL_ACCEPT_IPP
+#define BEAST_WEBSOCKET_IMPL_ACCEPT_IPP
+
+#include <beast/http/message.hpp>
+#include <beast/http/parser_v1.hpp>
+#include <beast/http/read.hpp>
+#include <beast/http/string_body.hpp>
```
Comment (vinniefalco) [inline include/beast/websocket/impl/accept.ipp]: `string_body.hpp`?  `http::string_body` is used a few lines down

File: include/beast/websocket/detail/stream_base.hpp (hunk 6 above)
Comment (vinniefalco) [inline include/beast/websocket/detail/stream_base.hpp]: willfix

File: doc/websocket.qbk (hunk 7 above)
Comment (vinniefalco) [inline doc/websocket.qbk]: More importantly, there is no `pong` function...willfix

File: include/beast/websocket/impl/ping.ipp [hunk 9]
```diff
@@ -229,6 +248,28 @@ ping(ping_data const& payload, error_code& ec)
     boost::asio::write(stream_, db.data(), ec);
 }
 
+template<class NextLayer>
+void
+stream<NextLayer>::
+pong(ping_data const& payload)
+{
+    error_code ec;
+    ping(payload, ec);
```
Comment (wilsonianb) [inline include/beast/websocket/impl/ping.ipp]: `ping` --> `pong`

Comment (vinniefalco) [inline include/beast/websocket/impl/ping.ipp]: oh lordy

**Comment by wilsonianb** [comment]
:+1: 
Just be sure to have the version commit at the end

**Comment by vinniefalco** [comment]
@wilsonianb That's a GitHub glitch when displaying pull requests
https://github.com/vinniefalco/Beast/commits/b18

**Pull request merged**